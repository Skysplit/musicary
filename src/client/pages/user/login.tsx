import React, { PureComponent } from 'react';
import { Card, CardText, CardTitle } from 'react-md';
import LoginFormContainer from '@app/client/containers/LoginFormContainer';

export default class UserLoginPage extends PureComponent {
  static async getInitialProps() {
    return { title: 'Login' };
  }

  render() {
    return (
      <Card>
        <CardTitle title="Login" />
        <CardText>
          <LoginFormContainer />
        </CardText>
      </Card>
    );
  }
}
