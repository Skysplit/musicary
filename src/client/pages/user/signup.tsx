import { PureComponent } from 'react';
import { Card, CardTitle, CardText } from 'react-md';
import SignupFormContainer from '@client/containers/SignupFormContainer';

export default class SignupPage extends PureComponent {
  static async getInitialProps() {
    return { title: 'Sign up!' };
  }

  render() {
    return (
      <Card>
        <CardTitle title="Sign up" />
        <CardText expandable={false}>
          <SignupFormContainer />
        </CardText>
      </Card>
    );
  }
}
