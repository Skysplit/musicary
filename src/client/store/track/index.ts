import { fetchTracksSuccess } from '@client/store/track/actions';
import { FETCH_MANY_TRACKS_SUCCESS } from '@client/store/track/actionTypes';

export interface TrackInterface {
  id: number;
  name: string;
  sourceId: string;
  sourceType: string;
}

export interface TrackState {
  byId: {
    [key: number]: {
      isLoading: boolean;
      isRemoving: boolean;
      data: TrackInterface;
    },
  };
}

type Actions =
  | ReturnType<typeof fetchTracksSuccess>;

const initialState = {
  byId: {},
};

export default (state: TrackState = initialState, action: Actions) => {
  switch (action.type) {
    case FETCH_MANY_TRACKS_SUCCESS: {
      return {
        ...state,
        byId: action.payload.tracks.reduce(
          (accum, track) => Object.assign(accum, {
            [track.id]: {
              isLoading: false,
              isRemoving: false,
              data: track,
            },
          }),
          {},
        ),
      };
    }

    default: {
      return state;
    }
  }
};
