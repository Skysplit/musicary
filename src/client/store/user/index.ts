import {
  fetchUserRequest,
  fetchUserSuccess,
  fetchUserFailure,
  logout,
} from '@client/store/user/actions';
import {
  FETCH_USER_SUCCESS,
  FETCH_USER_REQUEST,
  FETCH_USER_FAILURE,
  LOGOUT,
} from '@client/store/user/actionTypes';
import { getUserData, getUserToken } from '@client/utils/userData';

export interface UserInterface {
  id: number;
  email: string;
}

export interface UserState {
  user?: UserInterface;
  isLoading: boolean;
}

export type UserActions =
  | ReturnType<typeof fetchUserRequest>
  | ReturnType<typeof fetchUserSuccess>
  | ReturnType<typeof fetchUserFailure>
  | ReturnType<typeof logout>;


const canInitialize = global.window && getUserToken();

const initialState: UserState = {
  user: canInitialize ? getUserData() : null,
  isLoading: false,
};

export default (state: UserState = initialState, action: UserActions): UserState => {
  switch (action.type) {
    case FETCH_USER_REQUEST: {
      return {
        ...state,
        isLoading: true,
      };
    }

    case FETCH_USER_SUCCESS: {
      return {
        ...state,
        user: action.payload,
        isLoading: false,
      };
    }

    case FETCH_USER_FAILURE: {
      return {
        ...state,
        isLoading: false,
      };
    }

    case LOGOUT: {
      return {
        ...state,
        user: null,
      };
    }

    default: {
      return state;
    }
  }
};
