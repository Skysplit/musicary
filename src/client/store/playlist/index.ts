import { omit, map, uniq, reject } from 'lodash';
import {
  fetchPlaylistRequest,
  fetchPlaylistSuccess,
  fetchPlaylistsRequest,
  fetchPlaylistsSuccess,
  fetchPlaylistFailure,
  fetchPlaylistsFailure,
  setSinglePlaylist,
  removePlaylistRequest,
  removePlaylistSuccess,
  removePlaylistFailure,
  removePlaylistTrack,
} from '@client/store/playlist/actions';
import {
  FETCH_SINGLE_PLAYLIST_REQUEST,
  FETCH_SINGLE_PLAYLIST_SUCCESS,
  FETCH_MANY_PLAYLISTS_REQUEST,
  FETCH_MANY_PLAYLISTS_SUCCESS,
  FETCH_MANY_PLAYLISTS_FAILURE,
  FETCH_SINGLE_PLAYLIST_FAILURE,
  SET_PLAYLIST,
  REMOVE_PLAYLIST_REQUEST,
  REMOVE_PLAYLIST_SUCCESS,
  REMOVE_PLAYLIST_FAILURE,
  REMOVE_PLAYLIST_TRACK,
} from '@client/store/playlist/actionTypes';
import { fetchTracksSuccess, setSingleTrack } from '@client/store/track/actions';
import { FETCH_MANY_TRACKS_SUCCESS, SET_TRACK } from '@client/store/track/actionTypes';

export interface PlaylistInterface {
  id?: number;
  name: string;
  position?: number;
  tracks?: number[];
}

export interface PlaylistState {
  isLoading: boolean;
  byId: {
    [key: number]: {
      isLoading: boolean;
      isRemoving: boolean;
      data?: PlaylistInterface;
    };
  };
}

type Actions =
  | ReturnType<typeof fetchPlaylistRequest>
  | ReturnType<typeof fetchPlaylistSuccess>
  | ReturnType<typeof fetchPlaylistFailure>
  | ReturnType<typeof fetchPlaylistsRequest>
  | ReturnType<typeof fetchPlaylistsSuccess>
  | ReturnType<typeof fetchPlaylistsFailure>
  | ReturnType<typeof setSinglePlaylist>
  | ReturnType<typeof removePlaylistRequest>
  | ReturnType<typeof removePlaylistSuccess>
  | ReturnType<typeof removePlaylistFailure>
  | ReturnType<typeof fetchTracksSuccess>
  | ReturnType<typeof setSingleTrack>
  | ReturnType<typeof removePlaylistTrack>;


const initialState: PlaylistState = {
  isLoading: false,
  byId: {},
};

export default (state: PlaylistState = initialState, action: Actions): PlaylistState => {
  switch (action.type) {
    case FETCH_MANY_PLAYLISTS_REQUEST: {
      return {
        ...state,
        isLoading: true,
      };
    }

    case FETCH_MANY_PLAYLISTS_SUCCESS: {
      return {
        ...state,
        isLoading: false,
        byId: action.payload.reduce(
          (accum, data) => Object.assign(accum, {
            [data.id]: { data, isLoading: false },
          }),
          {},
        ),
      };
    }

    case FETCH_MANY_PLAYLISTS_FAILURE: {
      return state;
    }

    case FETCH_SINGLE_PLAYLIST_REQUEST: {
      return {
        ...state,
        byId: {
          ...state.byId,
          [action.payload]: {
            isLoading: true,
            isRemoving: false,
          },
        },
      };
    }

    case FETCH_SINGLE_PLAYLIST_SUCCESS: {
      return {
        ...state,
        byId: {
          ...state.byId,
          [action.payload.id]: {
            isLoading: false,
            isRemoving: false,
            data: action.payload,
          },
        },
      };
    }

    case FETCH_SINGLE_PLAYLIST_FAILURE: {
      return {
        ...state,
        byId: {
          ...state.byId,
          [action.payload]: {
            isLoading: false,
            isRemoving: false,
          },
        },
      };
    }

    case REMOVE_PLAYLIST_REQUEST: {
      const item = state.byId[action.payload];
      return {
        ...state,
        byId: {
          ...state.byId,
          [action.payload]: {
            ...item,
            isRemoving: true,
          },
        },
      };
    }

    case REMOVE_PLAYLIST_SUCCESS: {
      return {
        ...state,
        byId: omit(state.byId, [action.payload]),
      };
    }

    case REMOVE_PLAYLIST_FAILURE: {
      const item = state.byId[action.payload];
      return {
        ...state,
        byId: {
          ...state.byId,
          [action.payload]: {
            ...item,
            isRemoving: false,
          },
        },
      };
    }

    case SET_PLAYLIST: {
      return {
        ...state,
        byId: {
          ...state.byId,
          [action.payload.id]: {
            isLoading: false,
            isRemoving: false,
            data: action.payload,

          },
        },
      };
    }

    case FETCH_MANY_TRACKS_SUCCESS: {
      const { playlistId, tracks } = action.payload;
      const playlist = state.byId[playlistId];

      if (!playlist) {
        return state;
      }

      return {
        ...state,
        byId: {
          ...state.byId,
          [playlistId]: {
            ...playlist,
            data: {
              ...playlist.data,
              tracks: map(tracks, 'id'),
            },
          },
        },
      };
    }

    case SET_TRACK: {
      const { playlistId, track } = action.payload;
      const playlist = state.byId[playlistId];

      if (!playlist) {
        return;
      }

      return {
        ...state,
        byId: {
          ...state.byId,
          [playlistId]: {
            ...playlist,
            data: {
              ...playlist.data,
              tracks: uniq([
                ...playlist.data.tracks,
                track.id,
              ]),
            },
          },
        },
      };
    }

    case REMOVE_PLAYLIST_TRACK: {
      const { playlistId, trackId } = action.payload;
      const playlist = state.byId[playlistId];

      if (!playlist) {
        return;
      }

      return {
        ...state,
        byId: {
          ...state.byId,
          [playlistId]: {
            ...playlist,
            data: {
              ...playlist.data,
              tracks: reject(playlist.data.tracks, id => id === trackId),
            },
          },
        },
      };
    }

    default: {
      return state;
    }
  }
};
