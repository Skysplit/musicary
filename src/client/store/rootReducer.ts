
import { combineReducers } from 'redux';
import user from './user';
import playlist from './playlist';
import track from './track';

export default combineReducers({
  user,
  playlist,
  track,
});
