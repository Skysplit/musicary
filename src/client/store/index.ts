import { createStore } from 'redux';
import { StoreCreatorOptions } from 'next-redux-wrapper';
import { Request } from 'express';
import rootReducer from './rootReducer';
import middleware from './middleware';

export type State = ReturnType<typeof rootReducer>;

interface Options extends StoreCreatorOptions<State, {}, {}, {}, {}> {
  req: Request;
}

export default (initialState: State, options: Options) => {
  return createStore(rootReducer, middleware);
};
