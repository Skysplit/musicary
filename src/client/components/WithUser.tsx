import { PureComponent, ReactNode, ComponentClass, StatelessComponent } from 'react';
import { UserInterface } from '@client/store/user';
import { fetchUser } from '@app/client/store/user/operations';
import { getUserToken } from '@app/client/utils/userData';

export interface WithUserProps {
  user: UserInterface;
  loadingComponent: ComponentClass | StatelessComponent;
  isUserLoading: boolean;
  fetchUser: typeof fetchUser;
  forceFetch?: boolean;
  children: (user?: UserInterface) => ReactNode;
}

export default class WithUser extends PureComponent<WithUserProps> {
  static defaultProps = {
    forceFetch: false,
  };

  componentDidMount() {
    const { user, isUserLoading, forceFetch } = this.props;
    const token = getUserToken();

    if (!token) {
      return;
    }

    if ((isUserLoading || user) && !forceFetch) {
      return;
    }

    this.props.fetchUser();
  }

  render() {
    const {
      children,
      user,
      isUserLoading,
      loadingComponent: Loading,
    } = this.props;

    if (isUserLoading) {
      return <Loading />;
    }

    return children(user);
  }
}
