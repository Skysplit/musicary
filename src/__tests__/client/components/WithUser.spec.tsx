import React from 'react';
import { shallow } from 'enzyme';
import WithUser from '@client/components/WithUser';
import { UserInterface } from '@client/store/user';

describe('<WithUser />', () => {
  test('should render component', () => {
    const props = {
      user: { id: 1, email: 'test' } as UserInterface,
      children: jest.fn(user => <span>{user.id} {user.email}</span>),
    };
    const wrapper = shallow(<WithUser {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
