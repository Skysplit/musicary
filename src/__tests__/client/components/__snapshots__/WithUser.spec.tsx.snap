// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WithUser /> should render component 1`] = `
<span>
  1
   
  test
</span>
`;
